vllm/v1/core/sched/scheduler.py

```diff
@@ -1291,14 +1294,25 @@ def update_draft_token_ids(
                 # The request may have been finished. Skip.
                 continue
 
-            # Add newly generated spec token ids to the request.
+            # Filter out spec tokens which do not adhere to the grammar.
+            orig_num_spec_tokens = len(spec_token_ids)
             if self.structured_output_manager.should_advance(request):
                 metadata = request.structured_output_request
-                request.spec_token_ids = metadata.grammar.validate_tokens(  # type: ignore[union-attr]
-                    spec_token_ids
+                assert metadata is not None and metadata.grammar is not None
+                spec_token_ids = metadata.grammar.validate_tokens(
+                    spec_token_ids,
                 )
```

Comment:
nit: should fit now?
```suggestion
                spec_token_ids = metadata.grammar.validate_tokens(spec_token_ids)
```